test_conformance/subgroups/test_workgroup_non_uniform_arithmetic.cpp

```diff
@@ -0,0 +1,774 @@
+//
+// Copyright (c) 2020 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "procs.h"
+#include "subhelpers.h"
+#include "harness/typeWrappers.h"
+#include "workgroup_common_templates.h"
+
+
+// DESCRIPTION:
+// Test for scan inclusive non uniform functions
+// Which: 0 - add, 1 - max, 2 - min, 3 - mul, 4 - and, 5 - or, 6 - xor, 7 -
+// logical and, 8 - logical or, 9 - logical xor
+template <typename Ty, int Which> struct SCIN_NU
+{
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    {
+        genrand<Ty, Which>(x, t, m, ns, nw, ng);
+    }
+
+    static int chk(Ty *x, Ty *y, Ty *mx, Ty *my, cl_int *m, int ns, int nw,
```

Comment:
What are all these parameters?

Reply:
It was really confusing form me as well. Such code construction was from the beginning. I just used it. I've added comments inside of code to be able understand these names but I can write once again:

x - array with golden data calculated on host side
y - array with data read from the device (to compare)
mx - set of data extracted from x for given local workgoup
my - set of data extracted from y for given local workgoup
m - is an array which stores helpful information for test execution. It has values set by device (subgroup local id, subgroup id). Additionally it can contain different information set on host for subgroup_brodcast there is index of work_item which value will be broadcasted
nw - size of local work group
ng - number of local work groups
ns - size of subgroup
nj - number of subgroup

Reply:
I prefer adding such comment. Renaming variables should be as separate pull request.

Reply:
some improvements here in https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/d9fd232dd04e3ef1c479f33ad05204a329a0ab75  in code for ballot functions